Firestore/core/src/model/mutation.cc

```diff
@@ -107,79 +97,54 @@ bool Mutation::Rep::Equals(const Mutation::Rep& other) const {
          field_transforms_ == other.field_transforms_;
 }
 
-void Mutation::Rep::VerifyKeyMatches(
-    const absl::optional<MaybeDocument>& maybe_doc) const {
-  if (maybe_doc) {
-    HARD_ASSERT(maybe_doc->key() == key(),
-                "Can only apply a mutation to a document with the same key");
-  }
+void Mutation::Rep::VerifyKeyMatches(const MutableDocument& document) const {
+  HARD_ASSERT(document.key() == key(),
+              "Can only apply a mutation to a document with the same key");
 }
 
 SnapshotVersion Mutation::Rep::GetPostMutationVersion(
-    const absl::optional<MaybeDocument>& maybe_doc) {
-  if (maybe_doc && maybe_doc->type() == MaybeDocument::Type::Document) {
-    return maybe_doc->version();
+    const MutableDocument& document) {
+  if (document.is_found_document()) {
+    return document.version();
   } else {
     return SnapshotVersion::None();
   }
 }
 
-std::vector<FieldValue> Mutation::Rep::ServerTransformResults(
-    const absl::optional<MaybeDocument>& maybe_doc,
-    const std::vector<FieldValue>& server_transform_results) const {
-  HARD_ASSERT(field_transforms_.size() == server_transform_results.size(),
+TransformMap Mutation::Rep::ServerTransformResults(
+    const ObjectValue& previous_data,
+    const google_firestore_v1_ArrayValue& server_transform_results) const {
+  TransformMap transform_results;
+  HARD_ASSERT(field_transforms_.size() == server_transform_results.values_count,
               "server transform result size (%s) should match field transforms "
               "size (%s)",
-              server_transform_results.size(), field_transforms_.size());
+              server_transform_results.values_count, field_transforms_.size());
 
-  std::vector<FieldValue> transform_results;
-  for (size_t i = 0; i < server_transform_results.size(); i++) {
+  for (size_t i = 0; i < server_transform_results.values_count; i++) {
     const FieldTransform& field_transform = field_transforms_[i];
     const TransformOperation& transform = field_transform.transformation();
-
-    absl::optional<model::FieldValue> previous_value;
-    if (maybe_doc && maybe_doc->is_document()) {
-      previous_value = Document(*maybe_doc).field(field_transform.path());
-    }
-
-    transform_results.push_back(transform.ApplyToRemoteDocument(
-        previous_value, server_transform_results[i]));
+    const auto& previous_value = previous_data.Get(field_transform.path());
+    google_firestore_v1_Value transformed_value =
+        transform.ApplyToRemoteDocument(previous_value,
+                                        server_transform_results.values[i]);
+    transform_results.insert({field_transform.path(), transformed_value});
```

Comment:
It seems like in the iOS repo users could actually add more than one transform per field. They would have to use the update() API and update the same field multiple times, which always led to undefined behavior but is now "last writer wins". Using multiple transforms per field is no longer possible on Android/Web because of exactly this map and we should also reject it on iOS. I added a TODO to our input validation. FWIW, we "broke" this on the other platforms without anyone noticing.

Reply:
Cool, thanks for the explanation.